geomesa-jobs/src/main/scala/org/locationtech/geomesa/jobs/mapred/GeoMesaInputFormat.scala

```diff
@@ -0,0 +1,281 @@
+import org.opengis.filter.Filter
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions._
+import scala.collection.mutable.ArrayBuffer
+
+object GeoMesaInputFormat extends Logging {
+
+  /**
+   * Configure the input format.
+   *
+   * This is a single method, as we have to calculate several things to pass to the underlying
+   * AccumuloInputFormat, and there is not a good hook to indicate when the config is finished.
+   */
+  def configure(job: JobConf,
+                dsParams: Map[String, String],
+                featureTypeName: String,
+                filter: Option[String]): Unit = {
+
+    val ds = DataStoreFinder.getDataStore(dsParams).asInstanceOf[AccumuloDataStore]
+
+    assert(ds != null, "Invalid data store parameters")
+
+    // set up the underlying accumulo input format
+    val user = AccumuloDataStoreFactory.params.userParam.lookUp(dsParams).asInstanceOf[String]
+    val password = AccumuloDataStoreFactory.params.passwordParam.lookUp(dsParams).asInstanceOf[String]
+    InputFormatBase.setConnectorInfo(job, user, new PasswordToken(password.getBytes()))
+
+    val instance = AccumuloDataStoreFactory.params.instanceIdParam.lookUp(dsParams).asInstanceOf[String]
+    val zookeepers = AccumuloDataStoreFactory.params.zookeepersParam.lookUp(dsParams).asInstanceOf[String]
+    InputFormatBase.setZooKeeperInstance(job, instance, zookeepers)
+
+    val auths = Option(AccumuloDataStoreFactory.params.authsParam.lookUp(dsParams).asInstanceOf[String])
+    auths.foreach(a => InputFormatBase.setScanAuthorizations(job, new Authorizations(a.split(","): _*)))
+
+    // run an explain query to set up the iterators, ranges, etc
+    val ecql = filter.map(ECQL.toFilter).getOrElse(Filter.INCLUDE)
+    val query = new Query(featureTypeName, ecql)
+    val queryPlans = ds.explainQuery(featureTypeName, query, ExplainNull)
+
+    // see if the plan is something we can execute from a single table
+    val tryPlan = if (queryPlans.length > 1) None else queryPlans.headOption.filter {
+      case qp: JoinPlan => false
+      case _ => true
+    }
+
+    val queryPlan = tryPlan.getOrElse {
+      // this query has a join or requires multiple scans - instead, fall back to the ST index
+      logger.warn("Desired query plan requires multiple scans - falling back to spatio-temporal scan")
+      val sft = ds.getSchema(featureTypeName)
+      val featureEncoding = ds.getFeatureEncoding(sft)
+      val indexSchema = ds.getIndexSchemaFmt(featureTypeName)
+      val hints = ds.strategyHints(sft)
+      val version = ds.getGeomesaVersion(sft)
+      val queryPlanner = new QueryPlanner(sft, featureEncoding, indexSchema, ds, hints, version)
+      new STIdxStrategy().getQueryPlan(query, queryPlanner, ExplainNull)
+    }
+
+    // use the explain results to set the accumulo input format options
+    InputFormatBase.setInputTableName(job, queryPlan.table)
+    if (!queryPlan.ranges.isEmpty) {
```

Comment:
Yeah - I don't think that should ever happen. It would at least have a single range that spans the whole table. This is more a fail safe since I don't think you're supposed to pass empty ranges to InputFormatBase.